### PR TITLE
Create endpoint with pdf convertation user history

### PR DESCRIPTION
**Pull request by RezenkovD**

### CODE REVIEW

File: src/app/crud/crud_item.py [hunk 1]
```diff
@@ -1,9 +1,35 @@
...
+                and_(
+                    Item.user_id == user_id,
+                    extract("year", Item.time) == filter_date.year,
+                    extract("month", Item.time) == filter_date.month,
+                )
+            )
+            .all()
+        )
+    else:
+        return db.query(Item).filter_by(user_id=user_id).all()
```
Comment (qqlexa) [inline src/app/crud/crud_item.py:32]: Suggestion:
```python
db_query = db.query(Item).filter_by(user_id=user_id)
if filter_date is not None:
    db_query = db_query.filter(and_(year_condition, month_condition))
return db_query.all()
```

File: src/app/routers/pdf_convertations.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import jinja2
```
Comment (qqlexa) [inline src/app/routers/pdf_convertations.py:1]: Please check note comment about imports.

File: src/app/routers/pdf_convertations.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    return StreamingResponse(
+        task_result.result,
+        headers={"Content-Disposition": 'attachment; filename="my-history.pdf"'},
+        media_type="application/pdf",
+    )
+
+
+@router.post("/")
+async def pdf_convert(
+    filter_date: str = None,
```
Comment (qqlexa) [inline src/app/routers/pdf_convertations.py:38]: Optional[str]

Install mypy to check it automatically

File: src/app/routers/pdf_convertations.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+):
+    if filter_date is not None:
+        date_ = transform_date_or_422(filter_date)
+        data_user_stats = crud_statistics.get_user_statistics(
+            db, current_user.id, date_
+        )
+        list_items = read_items_for_user(db, current_user, date_)
+    else:
+        data_user_stats = crud_statistics.get_user_statistics(db, current_user.id)
+        list_items = read_items_for_user(db, current_user)
```
Comment (qqlexa) [inline src/app/routers/pdf_convertations.py:50]: I think we can do easier:
```python
if filter_date is not None:
    date_ = transform_date_or_422(filter_date)
else:
     date_ = None

data_user_stats = crud_statistics.get_user_statistics(db, current_user.id, date_)
list_items = read_items_for_user(db, current_user, date_)
```
Also, we can invert the condition:
```python
if filter_date is None:
    date_ = None
else:
     date_ = transform_date_or_422(filter_date)
```
Also, we can move `date_ = None` to begin.
```python
date_ = None
if filter_date:
     date_ = transform_date_or_422(filter_date)
```
What do you think?

File: src/app/templates/static/style.css [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+.btn.btn-white.disabled,
+.btn.btn-white.disabled:focus,
+.btn.btn-white.disabled:hover,
+.btn.btn-white[disabled],
+.btn.btn-white[disabled]:focus,
+.btn.btn-white[disabled]:hover {
+    color: #2d353c;
+    background: #fff;
+    border-color: #d9dfe3;
+}
```
Comment (qqlexa) [inline src/app/templates/static/style.css:132]: The better practice is to have an empty line at the end of a file.
https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file

File: src/app/templates/user-history-template.html [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                        {% endfor %}
+                     </tbody>
+                  </table>
+               </div>
+               <div><br></div>
+               <div class="invoice-price">
+                  <div class="invoice-price-left">
+                     <div class="invoice-price-row">
+                        <div class="sub-price">
+                           <span class="text-inverse">TOTAL PURCHASES : {{number_purchases}}</span><br>
```
Comment (qqlexa) [inline src/app/templates/user-history-template.html:82]: The variable should be between spaces: {{ number_purchases }}

File: src/app/utils/utils.py [hunk 7]
```diff
@@ -9,10 +9,11 @@
 
 SECRET_KEY = os.getenv("SECRET_KEY")
 ALGORITHM = "HS256"
-ACCESS_TOKEN_EXPIRE_MINUTES = 30
+ACCESS_TOKEN_EXPIRE_MINUTES = 3000
```
Comment (qqlexa) [inline src/app/utils/utils.py:12]: 3000 mins are 50 hours. It is intuitive to use 24 hours which equals a day or 48 hours which equals two days.

Comment (qqlexa) [reply src/app/utils/utils.py:12]: What about 2880 mins (48 hours)?

Comment (RezenkovD) [reply src/app/utils/utils.py:12]: You're right. 2880 will be clearer

File: src/app/crud/crud_item.py (hunk 1 above)
Comment (RezenkovD) [reply src/app/crud/crud_item.py:32]: In fact, it's the same :)

File: src/app/celery_app/celery_config.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import os
+
+from celery import Celery
+
+BACKEND = os.getenv("BACKEND")
+BROKER = os.getenv("BROKER")
+
+celery_app = Celery(
+    "main",
+    backend=BACKEND,
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/app/celery_app/celery_config.py:10]: Yes, we need it. 
[https://stackoverflow.com/questions/23215311/celery-with-rabbitmq-attributeerror-disabledbackend-object-has-no-attribute](url)

File: src/app/celery_app/tasks.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+
+@celery_app.task
+def pdf_convertation_user_history(output):
+    config = pdfkit.configuration(wkhtmltopdf=settings.WKHTMLTOPDF_PATH)
+    pdf_file = pdfkit.from_string(
+        output,
+        False,
+        configuration=config,
+        css="/home/riezienkov/projects/cost-accounting/src/app/templates/static/style.css",
```
Comment (qqlexa) [inline src/app/celery_app/tasks.py:15]: Env variable

File: src/app/crud/crud_item.py [hunk 10]
```diff
@@ -1,11 +1,37 @@
-from fastapi import HTTPException
+from fastapi import HTTPException, Depends
 from sqlalchemy.orm import Session
+from sqlalchemy import extract, and_
+from typing import Optional
```
Comment (qqlexa) [inline src/app/crud/crud_item.py:4]: Move it to begin of imports

File: src/app/crud/crud_item.py (hunk 1 above)
Comment (qqlexa) [reply src/app/crud/crud_item.py:32]: But you don't need to duplicate your code

File: src/app/routers/pdf_convertations.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
+from datetime import datetime
+
+from celery.result import AsyncResult
+from fastapi import Depends, APIRouter
+from sqlalchemy.orm import Session
+from starlette.responses import JSONResponse, StreamingResponse
+from starlette.templating import Jinja2Templates
+from typing import Optional
```
Comment (qqlexa) [inline src/app/routers/pdf_convertations.py:8]: Move it to top

File: src/app/routers/pdf_convertations.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    )
+    list_items = read_items_for_user(db, current_user, date_)
+
+    data_user_stats = data_user_stats.__dict__
+    today_date = {"today_date": datetime.today().strftime("%d %b, %Y")}
+
+    data_user_stats.update(today_date)
+    data_user_stats.update({"items": list_items})
+
+    template_loader = jinja2.FileSystemLoader("templates/")
```
Comment (qqlexa) [inline src/app/routers/pdf_convertations.py:58]: Why don't you use templates variable?

File: src/app/templates/user-history-template.html [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+                        {% for x in number_purchases_category %}
+                        <tr>
+                           <td>
+                              <span class="text-inverse">{{ x }}</span><br>
+                              <small></small>
+                           </td>
+                           <td class="text-center">{{ number_purchases_category[x] }}</td>
+                           <td class="text-right">{{ details[x] }} ₴</td>
+                        </tr>
+                        {%endfor%}
```
Comment (qqlexa) [inline src/app/templates/user-history-template.html:73]: Spaces between variables in {%%}

File: src/app/utils/utils.py [hunk 14]
```diff
@@ -1,18 +1,20 @@
 import os
+from datetime import datetime, date
+
 import logging
```
Comment (qqlexa) [inline src/app/utils/utils.py:4]: To top level imports

**Review by qqlexa — APPROVED** [review]

**Pull request merged**